Implement reusable "Add new site" component

yashwin opened the pull request:
<img width="1728" alt="Screenshot 2024-12-18 at 11 42 11 AM" src="https://github.com/user-attachments/assets/03314bc0-ab59-4427-88b9-d1fb11509093" />


<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/automattic-for-agencies-dev/issues/1574

## Proposed Changes

This PR implements a reusable "Add new site" component, which includes:

- All the components required.
- Context API to manage the toggle states for the modals.
- A4A menu items for "Add new site" flow.
- Async loading of env specific data

Note: 

- These changes are behind a feature flag.
- This PR enables this only on the dev & horizon environments. We will enable it on staging & production if everything looks good. 
- This PR implements the flow only on A4A. WPCOM changes will be made in a different PR.
- There are two flows where this was used, on the Overview and Sites dashboard. We have updated it in both places.
- I have kept the icons(WordPress & Pressable) the same as before to be consistent throughout the product. 

Fixing an issue I found here: https://github.com/Automattic/wp-calypso/pull/97636

## Why are these changes being made?

* To create a consistent experience of the "Add new site" flow

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Open the A4A live link below
* Go to the Overview page > Click the `Add sites` button on the top right > Verify the UI looks good on all the devices. Perform all the actions that are available and verify they work as expected. Best thing to do here is compare it with the production version of it by visiting https://agencies.automattic.com/

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

yashwin committed:
initial implementation

yashwin committed:
add feature flags and cleanup

yashwin committed:
ui enhancements

yashwin committed:
minor improvements

jkguidaven commented on client/components/add-new-site/index.tsx:
```diff
@@ -0,0 +1,64 @@
+	const renderPopoverContent = useMemo( () => {
+		switch ( true ) {
+			case isA8CForAgencies():
+				return <AddNewSiteMenuItems.A4A setMenuVisible={ setMenuVisible } />;
```
We should probably figure out how to asynchronously load this part, so we don't force non-A4A consumers to import code that is not relevant to their use case. 🤔  

Honestly, I'm not sure if there is any benefit in putting everything in one central place rather than letting each consumer (WPOM, Jetpack or A4A) build their own 'Add new site' flow using re-usable components.

Unless we have plenty of overlapping use cases, it would make sense to reuse the same logic and have everything centralized. 🤔

yashwin replied:
I have made the changes to load the modal asynchronously. Loading the menu items on load is a bit complicated in this case as the component gets rendered when the button is clicked. I think it's not a large piece of code, so it ok to load this synchronously as we need it to be available as soon as the button is clicked without compromising on the user experience.

On the other hand, I have implemented async loading for the modal here. So, that won't get loaded if unnecessary.

jkguidaven commented on client/components/add-new-site/button.tsx:
```diff
@@ -0,0 +1,43 @@
+			onClick={ toggleMenu }
+		>
+			<>
+				{ showMainButtonLabel ? translate( 'Add sites' ) : null }
+				<Gridicon
+					className={ clsx(
+						{ reverse: showMainButtonLabel && isMenuVisible },
+						{ mobile: ! showMainButtonLabel }
+					) }
+					icon={ showMainButtonLabel ? 'chevron-down' : 'plus' }
+				/>
```
Not necessarily required now, but it would be nice if we could override these in some way.

yashwin replied:
IMO, it is not required now as our goal is to keep the experience the same throughout.

jkguidaven commented on client/components/add-new-site/index.tsx:
```diff
@@ -0,0 +1,64 @@
+
+	return (
+		<AddNewSiteContext.Provider value={ { visibleModalType, setVisibleModalType } }>
+			<AddNewSiteButton
+				showMainButtonLabel={ ! isNarrowView }
+				isMenuVisible={ isMenuVisible }
+				toggleMenu={ toggleMenu }
+				popoverMenuContext={ popoverMenuContext }
+			/>
+			<AddNewSitePopover
+				isMenuVisible={ isMenuVisible }
+				toggleMenu={ toggleMenu }
+				popoverMenuContext={ popoverMenuContext }
+			>
+				{ renderPopoverContent }
+			</AddNewSitePopover>
+			{ renderModalsContent }
```
I believe the proposal was to implement the Compound component pattern. This appears to deviate from the initial plan based on the shaping document.. 🤔

yashwin replied:
If you look closely, we are still following the Compound component pattern; it's just that the components are rendered based on the env from this method. 

This method returns

```
<AddNewSitePopoverColumn>
  <AddNewSiteMenuItem />
</AddNewSitePopoverColumn>

jkguidaven replied:
But the `<AddNewSite>` didn’t have it, which makes a big difference. If we try to reuse this in a different context, we wouldn't be able to since we would have to create a new `<AddNewSite>`.

jkguidaven replied:
Ideally, we would want to make the `AddNewSite` agnostic for whatever content. But in this case, we will have to modify this if we want to add new content.

We can create a smart component that wraps the context logic (switcher) and then leave `AddNewSite` purely presentational.

yashwin replied:
> But the <AddNewSite> didn’t have it, which makes a big difference. If we try to reuse this in a different context, we wouldn't be able to since we would have to create a new <AddNewSite>.

Yes, that was the initial plan, but based on the feedback we got on shaping P2, we don't want to keep the component like that. 

```
For proper centralization and unification, I’d expect there would be just one use case for that component and it would be used for all environments, and any env-specific features would be determined inside it (with feature flags or something else similar).
```

jkguidaven replied:
I believe we could still implement the centralization and unification but still make the `AddNewSite` component agnostic and dumb. We can create a Wrapper component that builds menus based on the environment.

This makes the `AddNewSite` reusable in different contexts later. We have the flexibility to go with the unified approach or have each consumer build separately.

Here is an example of usage.
```
import AddNewSite from '..';

export function GlobalAddNewSiteButton() {
 	const renderPopoverContent = useMemo( () => {
		switch ( true ) {
			case isA8CForAgencies():
				return <AddNewSiteMenuItems.A4A setMenuVisible={ setMenuVisible } />;
			default:
				return null;
		}
	}, [ setMenuVisible ] );
 
 
  return <AddNewSite>{ renderPopoverContent }</AddNewSite>;
}
```

jkguidaven replied:
Anyway, if you do not anticipate any use cases that could benefit from the agnostic approach, we can proceed with this implementation.

yashwin replied:
I have made some changes to the approach here: https://github.com/Automattic/wp-calypso/blob/a70b714f55360175a7a0e16029d8df8de4b62ac0/client/components/add-new-site/index.tsx

yashwin committed:
refactor and use async loading for modals

yashwin commented:
Thanks for the review, @jkguidaven!

I have made some changes here. Async loading and some refactoring. Could you please take another look? Thanks again! 🙇

yashwin committed:
text changes

yashwin committed:
minor ui fix

nuriapenya commented:
Heya 👋🏼 cool work! small design review here:

* Should we use the official blue logo for WordPress.com? (Usually, the black version is reserved to Core)
* Could we add a bit more spacing between the list elements? they seem quite packed.
![screen 2024-12-18 at 09 40 04](https://github.com/user-attachments/assets/c1d47342-8031-4705-be9f-6b99563d3c1b)

cc. @lucasmendes-design

yashwin committed:
address PR comment

tyxla commented on client/components/add-new-site/index.tsx:
```diff
@@ -0,0 +1,43 @@
+
+import './style.scss';
+
+const AddNewSite = () => {
```
Thanks for centralizing this component so it's reusable by other environments easily 👍

tyxla commented on client/components/add-new-site/index.tsx:
```diff
@@ -0,0 +1,43 @@
+import { useRef, useState, useCallback } from 'react';
+import AddNewSiteButton from './button';
+import { AddNewSiteContext } from './context';
+import AddNewSiteMenuItems from './menu-items';
```
Any reason to go with a mixture of default and named exports?

yashwin replied:
Good catch! Removed [it now](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/index.tsx:
```diff
@@ -0,0 +1,43 @@
+	}, [] );
+
+	return (
+		<AddNewSiteContext.Provider value={ { visibleModalType, setVisibleModalType } }>
```
Does the value need memoization to avoid needless rerendering of the entire "add new site" tree?

yashwin replied:
Yes, it makes sense. Added [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-item.tsx:
```diff
@@ -0,0 +1,50 @@
+	isBanner,
+	disabled,
+	buttonProps,
+	extraContent,
```
Perhaps we could just use a `children` prop for this? It's much more ergonomic from the perspective of the component consumer.

yashwin replied:
Made the changes [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-item.tsx:
```diff
@@ -0,0 +1,50 @@
+	isBanner,
+	disabled,
+	buttonProps,
+	extraContent,
```
Usually we try to sort list of props alphabetically, as it makes it more readable for some folks.

yashwin replied:
Made the changes [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-item.tsx:
```diff
@@ -0,0 +1,50 @@
+	icon: JSX.Element;
+	iconClassName?: string;
+	heading: string;
+	description: string | TranslateResult;
```
Just `TranslateResult` should be enough, no?

```suggestion
	description: TranslateResult;
```

yashwin replied:
Yes. Fixed it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-item.tsx:
```diff
@@ -0,0 +1,50 @@
+				<Icon className="sidebar__menu-icon" icon={ icon } size={ ICON_SIZE } />
+			</div>
+			<div className="add-new-site__popover-button-content">
+				<div className="add-new-site__popover-button-heading">{ heading }</div>
+				<div className="add-new-site__popover-button-description">{ description }</div>
```
Should heading and/or description be optional?

yashwin replied:
I don't see any requirement for this. So, left it as-is.

tyxla replied:
Sounds good to me 👍

tyxla commented on client/components/add-new-site/menu-item.tsx:
```diff
@@ -0,0 +1,50 @@
+				'is-disabled': disabled,
+			} ) }
+		>
+			<div className={ clsx( 'add-new-site__popover-button-icon', iconClassName ) }>
```
Do we really need a wrapping div? Can't we do it with just some CSS?

yashwin replied:
It is needed the way it is currently implemented.

tyxla commented on client/components/add-new-site/menu-item.tsx:
```diff
@@ -0,0 +1,50 @@
+
+type Props = {
+	icon: JSX.Element;
+	iconClassName?: string;
```
Can we audit the PR for props we don't need? For example `iconClassName` seems to be unused. I'd always recommend the [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) as it helps us keep the codebase clean and simple.

yashwin replied:
Good catch. Removed it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-items/index.tsx:
```diff
@@ -0,0 +1,18 @@
+
+const AddNewSiteMenuItems = ( props: AddNewSiteMenuItemsProps ) => {
+	const renderContent = useMemo( () => {
+		switch ( true ) {
```
I find a `switch` unnecessary here - can't we just do a simple chain of `if` conditions? You don't even need to use `else` (although you can).

yashwin replied:
Made the changes [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-items/index.tsx:
```diff
@@ -0,0 +1,18 @@
+import AddNewSiteA4AMenuItems from './a4a';
+import type { AddNewSiteMenuItemsProps } from '../types';
+
+const AddNewSiteMenuItems = ( props: AddNewSiteMenuItemsProps ) => {
```
Thanks for abstracting the environment specifics by design 👍

tyxla commented on client/components/add-new-site/menu-items/index.tsx:
```diff
@@ -0,0 +1,18 @@
+			default:
+				return null;
+		}
+	}, [ props ] );
```
FWIW, I don't see a good reason to have a memo here. React's reconciliation algorithm should take care of it.

yashwin replied:
Made the changes [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-items/a4a.tsx:
```diff
@@ -0,0 +1,192 @@
+						onClick: () => {
+							setVisibleModalType( 'import-from-wpcom' );
+							setMenuVisible( false );
+						},
```
These `onClick` handlers are a bit repetitive, and just differ by the modal type. I wonder if we can create helper function to create them.

yashwin replied:
Made the changes [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-items/a4a.tsx:
```diff
@@ -0,0 +1,192 @@
+				/>
+				<AddNewSiteMenuItem
+					icon={ <A4ALogo /> }
+					heading={ translate( 'Via the Automattic plugin ' ) }
```
Let's audit the PR for leading/trailing spaces in strings, those can be problematic for translators sometimes

```suggestion
					heading={ translate( 'Via the Automattic plugin' ) }
```

yashwin replied:
Good catch! Fixed it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-items/a4a.tsx:
```diff
@@ -0,0 +1,192 @@
+			<AddNewSitePopoverColumn heading={ translate( 'Add a new production site' ) }>
+				<AddNewSiteMenuItem
+					icon={ <img src={ pressableIcon } alt="Pressable" /> }
+					heading={ translate( 'Pressable' ) }
```
Do we really want to translate `Pressable`?

yashwin replied:
Not needed. Removed it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/menu-items/a4a.tsx:
```diff
@@ -0,0 +1,192 @@
+				/>
+				<AddNewSiteMenuItem
+					icon={ <WordPressLogo /> }
+					heading={ translate( 'WordPress.com' ) }
```
Do we really want to translate `WordPress.com`

yashwin replied:
Not needed. Removed it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

nuriapenya commented:
Cool! The spaces look much better, thank you!

About the logo 🤔, I definitely don't want to introduce inconsistency. Let me align with designers and I'll get back to you asap. 🙏🏼

tyxla commented on client/components/add-new-site/modals/a4a/index.tsx:
```diff
@@ -0,0 +1,52 @@
+		getQueryArg( window.location.href, 'add_new_dev_site' )
+	);
+
+	useEffect( () => {
+		if ( shouldAutoOpenDevSiteConfigModal ) {
+			setVisibleModalType( 'dev-site-configurations' );
+		}
+	}, [ shouldAutoOpenDevSiteConfigModal, setVisibleModalType ] );
```
Looks like you may not need an effect here. Read [this](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) for more info.

yashwin replied:
I think it is still needed here. The other option we have is passing the state from the parent component, which I would like to avoid here as we are trying to keep the common components agnostic. What do you think?

yashwin replied:
Update:

I have removed this part as I feel this can be handled purely on A4A. Updated it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/b63fd4707d42c7d5463554caca73edc2ac851cf7).

tyxla replied:
Thanks. Note that this is a very common misuse of `useEffect`: adding an effect for conditionally setting state is most often not necessary. I can see plenty of cases across Calypso that can be cleaned up.

tyxla commented on client/components/add-new-site/modals/a4a/index.tsx:
```diff
@@ -0,0 +1,52 @@
+
+	const onCreateSiteSuccess = useSiteCreatedCallback( refetchRandomSiteName );
+
+	return (
+		<>
+			{ visibleModalType === 'a4a-connection' && <A4AConnectionModal onClose={ handleOnClose } /> }
+			{ visibleModalType === 'jetpack-connection' && (
+				<JetpackConnectionModal onClose={ handleOnClose } />
+			) }
+			{ visibleModalType === 'import-from-wpcom' && (
+				<ImportFromWPCOMModal onClose={ handleOnClose } />
+			) }
+			{ visibleModalType === 'dev-site-configurations' && (
+				<SiteConfigurationsModal
+					closeModal={ handleOnClose }
+					randomSiteName={ randomSiteName }
+					isRandomSiteNameLoading={ isRandomSiteNameLoading }
+					onCreateSiteSuccess={ onCreateSiteSuccess }
+				/>
+			) }
+		</>
+	);
```
Simple `if`s might be a bit more readable (untested):

```suggestion
	if ( visibleModalType === 'a4a-connection' ) {
		return <A4AConnectionModal onClose={ handleOnClose } />;
	}
	if ( visibleModalType === 'jetpack-connection' ) {
		return <JetpackConnectionModal onClose={ handleOnClose } />;
	}
	if ( visibleModalType === 'import-from-wpcom' ) {
		return <ImportFromWPCOMModal onClose={ handleOnClose } />;
	}
	if ( visibleModalType === 'dev-site-configurations' ) {
		return (
			<SiteConfigurationsModal
				closeModal={ handleOnClose }
				randomSiteName={ randomSiteName }
				isRandomSiteNameLoading={ isRandomSiteNameLoading }
				onCreateSiteSuccess={ onCreateSiteSuccess }
			/>
		);
	}
	return null;
```

yashwin replied:
Makes sense. Changed it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/modals/index.tsx:
```diff
@@ -0,0 +1,19 @@
+import isA8CForAgencies from 'calypso/lib/a8c-for-agencies/is-a8c-for-agencies';
+
+const AddNewSiteModals = () => {
+	const renderContent = useMemo( () => {
+		switch ( true ) {
+			case isA8CForAgencies():
+				return (
+					<AsyncLoad require="calypso/components/add-new-site/modals/a4a" placeholder={ null } />
+				);
+			default:
+				return null;
+		}
+	}, [] );
+	return renderContent;
```
As mentioned above, I'd skip the memoization here and would just go with an `if` and early return.

yashwin replied:
Makes sense. Changed it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

tyxla commented on client/components/add-new-site/modals/index.tsx:
```diff
@@ -0,0 +1,19 @@
+		switch ( true ) {
+			case isA8CForAgencies():
+				return (
+					<AsyncLoad require="calypso/components/add-new-site/modals/a4a" placeholder={ null } />
```
Optional: feel free to leave async loading for later, unless you're certain it will have a big impact.

yashwin replied:
~Yeah. I realized that this would require some thinking. So, I removed it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).~

I think we can keep this for now. What do you think? Updated [commit](https://github.com/Automattic/wp-calypso/pull/97499/commits/b63fd4707d42c7d5463554caca73edc2ac851cf7).

tyxla replied:
Yep, I agree - introducing async loading early can help address some of @jkguidaven's concerns.

tyxla reviewed (commented):
Thank you for the modular and environment-agnostic approach from the beginning 👏 

I think this is heading in a really good direction! 👍 

I've left a few suggestions and thoughts, most are minor things, let me know what you think.

jkguidaven commented on client/components/add-new-site/style.scss:
```diff
@@ -0,0 +1,166 @@
+			}
+
+			.add-new-site__popover-button-heading {
+				@include a4a-font-body-md($font-weight: 500);
```
I'm unsure if we should reference A4A-specific mixins in the shared scope. 🤔

yashwin replied:
Good catch! Fixed it [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

jkguidaven commented on client/components/add-new-site/menu-items/a4a.tsx:
```diff
@@ -0,0 +1,192 @@
+import { useTranslate } from 'i18n-calypso';
+import { useContext } from 'react';
+import A4ALogo from 'calypso/a8c-for-agencies/components/a4a-logo';
+import {
+	A4A_MARKETPLACE_HOSTING_PRESSABLE_LINK,
+	A4A_MARKETPLACE_HOSTING_WPCOM_LINK,
+	A4A_PAYMENT_METHODS_ADD_LINK,
+	A4A_SITES_LINK,
+	A4A_SITES_LINK_NEEDS_SETUP,
+} from 'calypso/a8c-for-agencies/components/sidebar-menu/lib/constants';
+import useFetchDevLicenses from 'calypso/a8c-for-agencies/data/purchases/use-fetch-dev-licenses';
+import useFetchPendingSites from 'calypso/a8c-for-agencies/data/sites/use-fetch-pending-sites';
+import usePressableOwnershipType from 'calypso/a8c-for-agencies/sections/marketplace/hosting-overview/hooks/use-pressable-ownership-type';
```
I'm a bit confused about how the dependency hierarchy looks like here.

A4A relies on AddNewSite, a globally shared component, but AddNewSite also depends on A4A for some of its logic, leading to a circular reference. Should we be concerned about code smell here, making this less maintainable in the future?

tyxla replied:
I may be misunderstanding, but I don't really see a circular reference here. From an architectural PoV, this makes sense because it provides us with a way to use this component across multiple environments. If it were an actual circular reference, we'd just move some code and fix it. Rather, it's a component that needs pieces of A4A for its logic, and that's perfectly fine, IMHO. What specifically makes it less maintainable?

jkguidaven replied:
I apologize if I wasn't clear enough about the circular reference. If you consider `client/a4a` as one module and `clients/components` as a separate module, you end up with two modules referencing each other. This makes it really hard to understand Calypso's dependency tree because we have no clear idea which module is above and which one is below.

tyxla replied:
Thanks for clarifying 👍 I don't see `client/a4a` as a regular module, though - it's packaged to be loaded as a section at the top level of the app, and you'll never really import it directly from another module. So I'm not really concerned about this being a circular reference. Yet again, we can always move things around if real circular reference concerns arise.

yashwin replied:
Yes, agree. I think we can think of moving it to `client/shared` or something similar if we face any issues! 👍

madebynoam reviewed (approved):
Lgtm! Thanks, Yashwin!

yashwin committed:
address PR comments

yashwin commented:
Thanks for the thorough review, @tyxla & @jkguidaven!

I have addressed all the comments & made some changes [here](https://github.com/Automattic/wp-calypso/pull/97499/commits/8365a989faeeda3c1d128f37813e2595c66f08de).

Could you please take another look? Thanks again!

jkguidaven commented:
I still strongly believe that a centralized approach is not right for us.

I don't see any strong benefit in merging Dotcom, Jetpack, and A4A implementations into one component. They don't have any overlapping use cases. 

What will happen here is that Dotcom will add another `if` condition and create its separate component (`<AddNewSiteDotcomMenuItems />`). The problem is that we don't reuse any implementation from Dotcom, and they don't use any implementation from us. We are just basically putting all implementations into one place, forcing everyone to load and import them all. 

```
const AddNewSiteMenuItems = ( props: AddNewSiteMenuItemsProps ) => {
	if ( isA8CForAgencies() ) {
		return <AddNewSiteA4AMenuItems { ...props } />;
	}

    if ( isDotcom() ) {
       return <AddNewSiteDotcomMenuItems />;
    }
	return null;
};
```

It is just equivalent to doing something like below but doing it in the same component. 
```
const DotComMenuItems = () => {
return <AddNewSite>
  <AddNewSiteDotComMenuItems />
</AddNewSite>;
}

const A4AAddNewSite = () => {
return <AddNewSite>
  </AddNewSiteA4AMenuItems />
</AddNewSite>;
}
```

yashwin committed:
minor refactor

yashwin committed:
async load env specific data

yashwin committed:
fix page refresh while redirecting

yashwin commented:
@jkguidaven:

Thanks for the feedback. I have updated the approach to load all the env-specific data in one-go. 

CC: @tyxla as you have previously reviewed the PR.

yashwin committed:
add comment

tyxla commented:
Thanks for all the hard work here 🙇 

I'll be doing another round of review now. 

> I still strongly believe that a centralized approach is not right for us.

I disagree here, there are good reasons for the centralization that I've listed in pfunGA-4cM-p2, plus there might be more overlap in the future than now, since the plan is to unify the environments and get them closer together.

Async loading is fine here since we only show a button initially, and all the popover + any additional business logic functionality can be conditionally loaded under feature flags and environment checks. For me this resolves most of the concerns around scaling and reusability and still allows all teams that use this component to work together under the same umbrella. While additional scalability and performance concerns can arise, I believe we can resolve them with proper encapsulation and modularization. 

I also wanted to say I really appreciate the work being done by the A4A team here, since it's laying the groundwork for all environments, and that will save everyone a bunch of time and effort 👏

tyxla commented on client/a8c-for-agencies/sections/sites/sites-header-actions/index.tsx:
```diff
@@ -48,13 +50,19 @@ export default function SitesHeaderActions( { onWPCOMImport }: Props ) {
+						toggleDevSiteConfigurationsModal={ toggleDevSiteConfigurationsModal }
+					/>
+				</div>
+			) }
```
At this point we could actually avoid the duplication and move the conditional logic inside the `div`, that way React will have one fewer element to treat:

```suggestion
			<div ref={ ( ref ) => setTourStepRef( ref ) }>
				{ isEnabled( 'a4a-updated-add-new-site' ) ? (
					<AddNewSite />
				) : (
					<AddNewSiteButton
						showMainButtonLabel={ ! isMobile }
						onWPCOMImport={ onWPCOMImport }
						toggleDevSiteConfigurationsModal={ toggleDevSiteConfigurationsModal }
					/>
				) }
			</div>
```

yashwin replied:
Sorry, I overlooked this one. Fixed it now: https://github.com/Automattic/wp-calypso/pull/97499/commits/fca51b48b4cb37a59940787dcfce45e901850041

tyxla commented on client/components/add-new-site/content/a4a/index.tsx:
```diff
@@ -0,0 +1,26 @@
+import AddNewSiteA4AMenuItems from '../../menu-items/a4a';
```
I'm not a proponent of relative paths because when moving code around, it may require extra work to fix them, and those are sometimes trickier to find without actually getting a build failure. But I don't have particularly strong feelings here since these may never be moved out of the "add new site" module.

yashwin replied:
I have now updated the paths: https://github.com/Automattic/wp-calypso/pull/97499/commits/fca51b48b4cb37a59940787dcfce45e901850041

tyxla reviewed (approved):
I've left a couple of little nits here and there, but this looks very substantial from an architectural PoV and works well in my testing 👍 

One thing that comes to mind is to add some documentation at the root of the "Add new site" component that explains our approach to modularity and environment agnosticism. We can then point to that document when adding new features to the component, especially ones that work in other environments. That should help us keep the component's features isolated and its architecture scalable. Such a doc can be added in another PR, no need to block this one. 

Thanks for the great work on this one @yashwin  🚀

yashwin committed:
address PR comments

yashwin commented:
Thank you, @tyxla! 🙇 

I have made the changes accordingly. I'll merge this as soon the tests pass!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/17069186).
